src/webutil/tableState.js

```diff
@@ -0,0 +1,244 @@
+      >Date</TableSortLabel>
+ */
+export type TableState<T> = {|
+  /**
+    A stateful representation of the current page data.
+   */
+  +currentPage: $ReadOnlyArray<T>,
+  /**
+    A stateful representation of the length of the filtered unpaginated data.
+   */
+  +length: number,
+  /**
+    A stateful representation of the current page index.
+   */
+  +pageIndex: number,
+  /**
+    A stateful representation of the number of rows per page.
+   */
+  +rowsPerPage: number,
+  /**
+    A stateful representation of the current sort name (the last sort name passed to setSortFn).
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+|};
+
+type FilterFns<T> = {[string]: null | ((T) => boolean)};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(
+  // The base array. It should ideally be memoized before being passed in.
+  data: Array<T>,
+  initialOptions?: {
+    // The initial rows per page. Defaults to unpaginated.
+    initialRowsPerPage?: number,
+    // Defaults to unsorted (maintains order of the original data).
+    initialSort?: {
+      // The starting order of the initial sort, represented by "asc" or "desc".
+      sortOrder: SortOrderTypes,
+      // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+      sortName: string,
+      // A function that returns the property of the element that should be used in the sort comparator.
+      sortFn: PluckFn<T>,
+    },
+  }
+): TableState<T> {
+  const optionDefaults = {
+    initialRowsPerPage: data.length,
```

Comment:
It's kind of a code smell to me that the default is dependent on the data. I imagine it could result in issues where e.g. the data is set to have 5 rows, and then the data is updated to have 6 rows, and now the table is only showing the first 5 rows, resulting in user confusion.

That said: do we actually want the default to be "show all data"? Since this is UI code, we can imagine cases where' that's quite bad, e.g. it tries to render 20,000 rows and hangs. Maybe go with a safer default like 25?

Reply:
This issue is that it is the responsibility of the UI view to create pagination controls, and the TableState cannot coerce the UI view to use pagination. I prefer the unidirectional semantics of TableState offering Sort, Filter, and Pagination and the UI view cherry-picking which it needs.